woocommerce/google-listings-and-ads
# Add/1165 request modal

### Changes proposed in this Pull Request

Closes part of #1165 

The intention of this PR is to add the Request Notice and Request Confirmation Modal functionality in regards [GLA Request Review](pcTzPl-bx-p2) as well as a basic scaffolding for the API endpoint necessary to get the Account Status Information.



⚠️ Notice the Google API is not ready yet. (TBD by Google). And we are hardcoding dummy responses in order to simulate the API and continue working on the frontend part until the API is done.  Therefore this code is potentially going to change significantly after the API is deployed. This PR is not implementing the new issues fetcher. It only implements the notice, the modal, and the issues to show in the modal... not the issues in the table that are still working with the old version. So, it is ok to see different issues in the modal and the Account Issues table. 

### Changes added

- Dummy API & Scaffolding in the backend
- Request Review Notice functionality in the frontend
- Request Review Modal functionality in the frontend
- Basic tests 

### Screenshots:

<img width="1062" alt="Screenshot 2022-02-08 at 10 30 45" src="https://user-images.githubusercontent.com/5908855/152931650-591461a3-7f1e-41a3-8f58-fadcb900620b.png">

<img width="771" alt="Screenshot 2022-02-08 at 10 30 57" src="https://user-images.githubusercontent.com/5908855/152931679-794827d0-72c6-4240-9caa-6d44ca90b83c.png">

### Detailed test instructions:

*For all the states*

1.  Install GLA and connect your account. (It's ok if you test with a previous connected Account) 
2. Go to the Product feed section. It's important that you have at least one issue Product or Account. Otherwise the issues block won't appear. So, in case you don't have issues you can do so by creating one issue in the `wp_gla_merchant_issues` table from the database 
3. Click on the Account Issues tab.

*DISAPPROVED STATE*

1. You will see this in the notice

<img width="1062" alt="Screenshot 2022-02-08 at 10 30 45" src="https://user-images.githubusercontent.com/5908855/152932620-2e09189e-cdf7-4c57-80fb-9630a541becb.png">

2. You can click on Request Review 
3. It will open a modal like this

<img width="771" alt="Screenshot 2022-02-08 at 10 30 57" src="https://user-images.githubusercontent.com/5908855/152932687-cd771fa0-fe0d-4ef4-8555-08e903bf3590.png">

4. By default you only see 5 issues and a button to show 1 issue more (since there are 6 issues in the response)
5. Clicking on `+1 more issues ` will show the rest of the issues (in this case 1 more, the issue 6)
6. Request review button is disabled
7. If you check the `I have resolved all the issue(s) listed above.` the button is enabled.
8. You can close the modal by clicking cancel, by clicking on the X at the corner or by clicking outside the modal
9. You can click again on the `Request review` button in the notice and it opens the modal

10. Go to `src/API/Site/Controllers/MerchantCenter/AccountReviewController.php`  in the code
11. In Line 74 add another issue. `#7 issue seven` 
12. Go to the admin again, refresh, open the modal and see that still only 5 issues are shown
13. Check that now it appears `+2 more issues `
14. Click on that and see how issues 6 and 7 are shown.  And text changes to "Show less" 
15. Click again and see how 6 and 7 disappear. And text is again `+2 more issues `

16. Go to `src/API/Site/Controllers/MerchantCenter/AccountReviewController.php`  in the code
17. In Line 74 remove some issues in a way you have only 5 issues
18. Go to the admin again, refresh, open the modal and see that still only 5 issues are shown
19. Show more button is not visible anymore

*WARNING STATE*

1. Go to `src/API/Site/Controllers/MerchantCenter/AccountReviewController.php`  in the code
2. In L66 change the line to ->  `'status'  => $this->request_review->is_allowed() ? RequestReview::REVIEW_STATUS_WARNING : RequestReview::REVIEW_STATUS_BLOCKED,`
3. Refresh the admin page. 
4. See how the same notice as the DISAPPROVED state appears. And you can open the modal.

*BLOCKED STATE*
1. Go to Database and create a new row in `wp_options` table 

```
Option name: `gla_mc_next_review_requested_at`
Option value:  Unix timestamp at some point in the future. You can use `1675814400` (8 feb 2023) or create your own using https://www.unixtimestamp.com/ (notice it should be created at 00.00.00 UTC of the desired day)
```

2. Refresh the admin
3. See the notice like this
<img width="964" alt="Screenshot 2022-02-08 at 11 01 43" src="https://user-images.githubusercontent.com/5908855/152935039-9c55c185-0320-400f-aec3-57ad19ae7129.png">

4. Be sure the date appears and makes sense with the timestamp inserted in the `wp_option`
5. Learn more link appears and works
6. Review button is disabled and non functional

7. Go back to Database and edit the option with a timestamp in the past... for example `1612742400` (8 feb 2021)
8. Refresh the admin. Date info disappeared, and Review Button is enabled again.


*PENDING STATE*

1. Go to `src/API/Site/Controllers/MerchantCenter/AccountReviewController.php`  in the code
2. In L66 change the line to ->  `'status'      => RequestReview::REVIEW_STATUS_UNDER_REVIEW,`
3. Refresh the admin page. 
4. See how this notice appears
<img width="968" alt="Screenshot 2022-02-08 at 11 07 08" src="https://user-images.githubusercontent.com/5908855/152935725-2f16fb49-5d65-4af8-8ca1-2ba24ddfdd49.png">

*APPROVED STATE*

1. Go to `src/API/Site/Controllers/MerchantCenter/AccountReviewController.php`  in the code
2. In L66 change the line to ->  `'status'      => RequestReview::REVIEW_STATUS_APPROVED,`
3. Refresh the admin page. 
4. See how no notice appears anymore


### Changelog entry

>
